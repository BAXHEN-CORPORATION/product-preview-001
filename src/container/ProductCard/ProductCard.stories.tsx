import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";

import img from "../../assets/product_image.png";

import ProductCard from "./ProductCard";
import Stack from "@mui/material/Stack";

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: "Containers/ProductCard",
  component: ProductCard,
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {
    title: {
      control: "text",
      description: "Product's name",
    },
    category: {
      control: "text",
      description: "Product's category (i.e PERFUME, ELECTRONICS, FASHION)",
    },
    description: { control: "text", description: "Product's description" },
    price: { control: "number", description: "Product's end price" },
    fullPrice: {
      control: "number",
      description: "Product's full price without discount",
    },
    productImageUrl: {
      control: "text",
      description: "Product's photo",
      defaultValue:
        "https://upload.wikimedia.org/wikipedia/commons/thumb/6/65/No-Image-Placeholder.svg/1665px-No-Image-Placeholder.svg.png",
    },
    onClick: { control: "none" },
  },
} as ComponentMeta<typeof ProductCard>;

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof ProductCard> = (args) => (
  <Stack
    alignItems="center"
    justifyContent="center"
    bgcolor="common.cream"
    p={8}
  >
    <ProductCard {...args} />
  </Stack>
);

export const DefaultProductCard = Template.bind({});
// More on args: https://storybook.js.org/docs/react/writing-stories/args
DefaultProductCard.args = {
  title: "Gabrielle Essence Eau De Parfum",
  category: "PERFUME",
  description:
    "A floral, solar and voluptuous interpretation composed by Olivier Polge, Perfumer-Creator for the House of CHANEL.",
  fullPrice: 169.99,
  price: 149.99,
  productImageUrl: img,
};
export const NoImageProductCard = Template.bind({});
// More on args: https://storybook.js.org/docs/react/writing-stories/args
NoImageProductCard.args = {
  title: "Gabrielle Essence Eau De Parfum",
  category: "PERFUME",
  description:
    "A floral, solar and voluptuous interpretation composed by Olivier Polge, Perfumer-Creator for the House of CHANEL.",
  fullPrice: 169.99,
  price: 149.99,
};
