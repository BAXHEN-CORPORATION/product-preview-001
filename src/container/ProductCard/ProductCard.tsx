//** External Imports */
import React from "react";

import Stack from "@mui/material/Stack";
import Box from "@mui/material/Box";
import Typography from "@mui/material/Typography";
import Button from "@mui/material/Button";
import CartIcon from "../../components/CartIcon";

//** Local Imports */

//** Typings */
export interface ProductCardProps {
  category: string;
  title: string;
  description: string;
  price: number;
  discountPrice: number;
  productUrl: string;
}

//** Default Props */
const defaultProps: Partial<ProductCardProps> = {};

/**
 * Container for displaying product info
 * @params category
 * @params title
 * @params description
 * @params price
 * @params discountPrice
 * @params productUrl
 *
 * @container
 */
const ProductCard: React.FC<ProductCardProps> = ({
  productUrl,
  title,
  category,
  description,
  price,
  discountPrice,
}) => {
  return (
    <Stack
      sx={{
        height: { mobile: "100%", tablet: "450px" },
        width: { mobile: "100%", tablet: "600px" },
        borderRadius: "10px",
        overflow: "hidden",
      }}
      direction={{ mobile: "column", tablet: "row" }}
    >
      <Box
        height={{ mobile: "240px", tablet: "100%" }}
        flex={{ mobile: "unset", tablet: 1 }}
      >
        <img
          style={{ width: "100%", height: "100%", objectFit: "cover" }}
          src={productUrl}
          alt={title}
        />
      </Box>
      <Stack flex={1} bgcolor="common.white" padding="24px" spacing={3}>
        <Typography variant="overline">{category}</Typography>
        <Typography variant="display">{title}</Typography>
        <Typography variant="body">{description}</Typography>

        <Stack direction="row" spacing="19px">
          <Typography variant="display" color="primary.main">
            ${discountPrice}
          </Typography>
          <Typography variant="body" component="del">
            ${price}
          </Typography>
        </Stack>

        <Button variant="contained" startIcon={<CartIcon />}>
          Add to Cart
        </Button>
      </Stack>
    </Stack>
  );
};

ProductCard.defaultProps = defaultProps;

export default ProductCard;
