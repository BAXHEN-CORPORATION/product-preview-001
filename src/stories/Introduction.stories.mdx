import { Meta } from "@storybook/addon-docs";

<Meta title="Introduction" />

# Introduction to Product Preview Card
