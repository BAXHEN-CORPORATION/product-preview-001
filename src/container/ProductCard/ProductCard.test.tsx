import { render, screen } from "@testing-library/react";
import { ProductCardProps, productImageUrl } from "./ProductCard";

import { DefaultProductCard, NoImageProductCard } from "./ProductCard.stories";

test("should render product card", () => {
  const { container } = render(
    <DefaultProductCard {...(DefaultProductCard.args as ProductCardProps)} />
  );

  expect(container.firstChild).toMatchSnapshot();
});

test("should render product card default props", () => {
  const onClick = jest.fn();

  render(
    <DefaultProductCard
      {...(DefaultProductCard.args as ProductCardProps)}
      onClick={onClick}
    />
  );

  const productImg = screen.getByTestId("image");

  const category = screen.getByTestId("category");
  const title = screen.getByTestId("title");
  const description = screen.getByTestId("description");
  const price = screen.getByTestId("price");
  const fullPrice = screen.getByTestId("full-price");
  const btn = screen.getByTestId("btn");

  expect(productImg.getAttribute("src")).toBe(
    DefaultProductCard.args?.productImageUrl
  );

  expect(category.textContent).toBe(DefaultProductCard.args?.category);
  expect(title.textContent).toBe(DefaultProductCard.args?.title);
  expect(description.textContent).toBe(DefaultProductCard.args?.description);
  expect(price.textContent).toBe(`$${DefaultProductCard.args?.price}`);
  expect(fullPrice.textContent).toBe(`$${DefaultProductCard.args?.fullPrice}`);
  expect(btn.textContent).toBe("Add to Cart");

  btn.click();

  expect(onClick).toBeCalled();
});

test("should render product card default props", () => {
  render(
    <NoImageProductCard {...(NoImageProductCard.args as ProductCardProps)} />
  );

  const productImg = screen.getByTestId("image");

  expect(productImg.getAttribute("src")).toBe(productImageUrl);
});
