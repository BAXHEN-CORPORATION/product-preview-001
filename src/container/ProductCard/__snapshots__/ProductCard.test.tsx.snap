// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render product card 1`] = `
<div
  class="css-v2lj72-MuiStack-root"
>
  <div
    class="css-1jgsxbd-MuiStack-root"
  >
    <div
      class="MuiBox-root css-11wvxpf"
    >
      <img
        alt="Gabrielle Essence Eau De Parfum"
        data-testid="image"
        src="product_image.png"
        style="width: 100%; height: 100%; object-fit: cover;"
      />
    </div>
    <div
      class="css-7u5uaq-MuiStack-root"
    >
      <span
        class="MuiTypography-root MuiTypography-overline css-b3fch5-MuiTypography-root"
        data-testid="category"
      >
        PERFUME
      </span>
      <span
        class="MuiTypography-root MuiTypography-display css-37jdci-MuiTypography-root"
        data-testid="title"
      >
        Gabrielle Essence Eau De Parfum
      </span>
      <span
        class="MuiTypography-root MuiTypography-body css-37jdci-MuiTypography-root"
        data-testid="description"
      >
        A floral, solar and voluptuous interpretation composed by Olivier Polge, Perfumer-Creator for the House of CHANEL.
      </span>
      <div
        class="css-t2p4tw-MuiStack-root"
      >
        <span
          class="MuiTypography-root MuiTypography-display css-1kbbegv-MuiTypography-root"
          data-testid="price"
        >
          $
          149.99
        </span>
        <del
          class="MuiTypography-root MuiTypography-body css-37jdci-MuiTypography-root"
          data-testid="full-price"
        >
          $
          169.99
        </del>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        data-testid="btn"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            fill="white"
            height="16"
            viewBox="0 0 15 16"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M12.8649 8.1664C13.7779 8.52626 14.3794 9.40635 14.3831 10.3878C14.381 11.2088 13.9589 11.9716 13.2644 12.4094C13.4723 12.7661 13.5825 13.1712 13.584 13.584C13.5914 14.5429 13.0268 15.4139 12.1485 15.7985C11.2701 16.1831 10.2471 16.0073 9.54753 15.3516C8.84795 14.6959 8.60643 13.6863 8.93348 12.785H7.0477C7.40887 13.78 7.07458 14.8942 6.22528 15.5261C5.37597 16.158 4.21274 16.158 3.36344 15.5261C2.51413 14.8942 2.17985 13.78 2.54101 12.785H1.59812C1.15681 12.785 0.799059 12.4272 0.799059 11.9859C0.799059 11.5446 1.15681 11.1868 1.59812 11.1868H11.9859C12.4272 11.1868 12.785 10.8291 12.785 10.3878C12.785 9.94648 12.4272 9.58873 11.9859 9.58873H3.76357C3.39211 9.60145 3.06093 9.35638 2.96451 8.99742L0.982843 1.59813H0V1.548e-05H1.59812C1.96072 -0.00157591 2.27893 0.241186 2.37321 0.591319L2.6369 1.59813H13.584C13.8259 1.59095 14.0581 1.69379 14.2153 1.87781C14.3692 2.07435 14.4224 2.33151 14.3591 2.57299L12.8649 8.1664ZM4.39483 7.99061H11.2587L12.5452 3.19625H3.06839L4.39483 7.99061ZM4.79436 14.3831C4.35305 14.3831 3.9953 14.0253 3.9953 13.584C3.9953 13.1427 4.35305 12.785 4.79436 12.785C5.23566 12.785 5.59342 13.1427 5.59342 13.584C5.59342 14.0253 5.23566 14.3831 4.79436 14.3831ZM10.3878 13.584C10.3878 14.0253 10.7455 14.3831 11.1868 14.3831C11.6281 14.3831 11.9859 14.0253 11.9859 13.584C11.9859 13.1427 11.6281 12.785 11.1868 12.785C10.7455 12.785 10.3878 13.1427 10.3878 13.584Z"
              data-testid="cart-icon"
              fill="white"
              fill-rule="evenodd"
            />
          </svg>
        </span>
        Add to Cart
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
